### PR TITLE
Add backend pagination for resource and party lists

### LINKED ISSUE #1465: Add backend pagination for listing resources
Listing a large number of resources contained within a project could result in having performance issues as there is no backend pagination implemented. We should implement backend pagination for listing resources to avoid this.

### LINKED ISSUE #1466: Add backend pagination for listing parties
Listing a large number of parties contained within a project (with the parties subnav) could result in having performance issues as there is no backend pagination implemented. We should implement backend pagination for listing parties to avoid this.

### PR DESCRIPTION
**Pull request by oliverroick**
### Proposed changes in this pull request

- Fixes #1465, add backend pagination to list-resources pages for project resources, location detail, relationship detail and party detail as well as lists for attaching existing resources to projects, locations, relationships, and parties. 
- Fixes #1466: Adds backend pagination to the list-party page. 

#### General idea

Instead of rendering all results into a page and have data tables slice it up into pages, we have data tables send requests to a particular asynchronous API endpoint. The endpoint takes care of rendering the current page using only the objects the user currently views, which reduces the amount of data we have to query from the data base, render and send to the client. 

Data tables sends requests in this form: `someurl?search[value]={string}&order[0][column]={int}&order[0][dir]={'asc'|'desc'}&start={int}&length={int}`.

It expects a response like: 

```javascript
{
    "recordsTotal": 10  // total number of records
    "recordsFiltered": 10  // number of filtered records (e.g. that match the search query
    "data": []  // all the data to render
}
```

Since we're rendering the tables on the server-side, we'll leave the `data` field empty and add a `tbody` field instead that contains the rendered table content. Upon receiving the response, we insert the content of `tbody` to the table's `tbody` element without further processing. 

_NOTE: This is obviously a workaround, it's neither a good solution nor is it particularly beautiful. To be honest, I shudder every time I look at it. But it's the best we can get given the technical constraints we're facing. For the future, we should consider serving information loaded asynchronously onto a page using real JSON-powered API endpoints and using a JS library like React to render the contents onto the page. (Something we should adapt soon, IMO.)_

#### New views

- `core.views.mixins.AsyncList`: Implements filtering and sorting functionality used in all views that follow below. It takes the view's query set then applies filtering for a search and order; and sliced the query set given the current page and size of the page. 
- `party.views.async.PartyList`: List all parties in a project. The view returns a rendered HTML table of the parties.
- `resources.views.async.ResourcesViews`: Abstracts functional used in both `ResourceList` and `ResourcAdd`; namely getting the correct entity object and building the search query. 
- `resources.views.async.ResourceList`: List all resources attached to an entity (project, party, location or relationship). The entity type is identified via the `contect_object` view argument; it defaults to `project.Project`  and is otherwise defined via the URL config. The template to render the list of resources is identified via the `template` view argument; it defaults to `resources/table_snippets/resource.html`  and is otherwise defined via the URL config. The view, returns a rendered HTML table of the parties.
- `resources.views.async.ResourceAdd`: Lists all resources that can be added to an entity, i.e. all resources that are uploaded to a project but not attached to the entity. Entity type and template are identified in the same way as in `ResourceList`. 

#### Changed views

List views have been changed so that the list of objects is not added to the view's template context. (The list of objects is loaded asynchronously using datatables).

Views affected are: 

- `resources.views.default.ProjectResources`
- `spatial.views.default.LocationDetail`
- `party.views.default.PartyDetail`
- `party.views.default.PartyRelationshipDetail`

#### Removed code

- `resources.fields.*` and `resources.forms.AddResourceFromLibrary`; resources are no longer added via a form
- `resources.views.mixins.DetachableResourcesListMixin`. This mixin found all resources that are attached to the current entity. Since all resources listings are generated in `resources.views.async.ResourcesList`, this functionality has been folded in that view (L78-85).

### When should this PR be merged

For the next release

### Risks

Low, none foreseen in particular

### Follow-up actions

- We should consider implementing a cleaner solution using a JSON API and front-end rendering.
- The least we should do is to apply the same approach to all other paginated views. 

#### General

- **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description. 
  - [ ] Review 1
  - [ ] Review 2
- **Is the PR labeled correctly?** It should have the `migration` label if a new migration is added. 
  - [ ] Review 1
  - [ ] Review 2
- **Is the risk level assessment sufficient?** The risks section should contain all risks that might be introduced with the PR and which actions we need to take to mitigate these risks. Possible risks are database migrations, new libraries that need to be installed or changes to deployment scripts.
  - [ ] Review 1
  - [ ] Review 2

#### Functionality

- **Are all requirements met?** Compare implemented functionality with the requirements specification.
  - [ ] Review 1
  - [ ] Review 2
- **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled. 
  - [ ] Review 1
  - [ ] Review 2

#### Code

- **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
  - [ ] Review 1
  - [ ] Review 2
- **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests. 
  - [ ] Review 1
  - [ ] Review 2
- **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/). 
  - [ ] Review 1
  - [ ] Review 2
- **Is the code documented sufficiently?** Large and complex classes, functions or methods must be annotated with comments following our [code-style guidelines](https://devwiki.corp.cadasta.org/Contributing%20Style%20Guide#documentation-and-comments).
  - [ ] Review 1
  - [ ] Review 2
- **Has the scalability of this change been evaluated?**
  - [ ] Review 1
  - [ ] Review 2
- **Is there a maintenance plan in place?**
  - [ ] Review 1
  - [ ] Review 2

#### Tests

- **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components. 
  - [ ] Review 1
  - [ ] Review 2
- **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
  - [ ] Review 1
  - [ ] Review 2
- **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?
  - [ ] Review 1
  - [ ] Review 2

#### Security

- **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
  - [ ] Review 1
  - [ ] Review 2
- **Are all UI and API inputs run through forms or serializers?** 
  - [ ] Review 1
  - [ ] Review 2
- **Are all external inputs validated and sanitized appropriately?**
  - [ ] Review 1
  - [ ] Review 2
- **Does all branching logic have a default case?**
  - [ ] Review 1
  - [ ] Review 2
- **Does this solution handle outliers and edge cases gracefully?**
  - [ ] Review 1
  - [ ] Review 2
- **Are all external communications secured and restricted to SSL?**
  - [ ] Review 1
  - [ ] Review 2

#### Documentation

- **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes must be documented in the [Cadasta Platform Documentation](https://github.com/Cadasta/cadasta-docs).
  - [ ] Review 1
  - [ ] Review 2
- **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented in the [API docs](https://github.com/Cadasta/api-docs).
  - [ ] Review 1
  - [ ] Review 2
- **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki. 
  - [ ] Review 1
  - [ ] Review 2

### CODE REVIEW

File: cadasta/templates/party/party_list.html [hunk 1]
```diff
@@ -15,24 +15,64 @@
...
   </div>
 </div>
 
+<script>
+window.addEventListener('load', function() {
+  $('#party-list').DataTable({
+    serverSide: true,
+    drawCallback: function (settings) {
+      var response = this.api().ajax.json().tbody;
+      $('#party-list tbody').html(this.api().ajax.json().tbody);
```
Comment (bjohare) [inline cadasta/templates/party/party_list.html]: Using the `response` variable to set the html here would be more readable.

File: cadasta/templates/resources/table_snippets/datatables.html [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    {% if large_table %}
+    columnDefs: [{
+      targets: 4,
+      orderable: false
+    }],
+    {% endif %}
+    drawCallback: function (settings) {
+      var dataTable = this;
+      var response = dataTable.api().ajax.json().tbody;
+      $(dataTable).find('tbody').html(dataTable.api().ajax.json().tbody);
```
Comment (bjohare) [inline cadasta/templates/resources/table_snippets/datatables.html]: Same here..

**Comment by bjohare** [comment]
Screenshot of hidden `attach` button during async content loading...

attach

File: cadasta/templates/party/party_list.html (hunk 1 above)
Comment (oliverroick) [inline cadasta/templates/party/party_list.html]: I have removed `var response = this.api().ajax.json().tbody;`; no need to store the response.

File: cadasta/templates/resources/table_snippets/datatables.html (hunk 2 above)
Comment (oliverroick) [inline cadasta/templates/resources/table_snippets/datatables.html]: Same here

**Comment by alukach** [comment]
I'm going to go ahead and be that person who asks the question: 

> "Why don't we just do this with the front-end JS framework that we all agree we should use?"

Some possible responses to this question (and my retorts):

> "Because we've already built this without the front-end JS framework. Let's use that next time, down the road."

 This PR currently has 46 file changes, with a net addition of 1,603 lines of code. It adds `/async` endpoints, which I believe to be an anti-pattern as it is a distraction from developing our API. We all agree that this isn't an ideal approach. I feel like this is adding a chunk of technical debt to our system.

> "We're not really familiar with React.js/Angular2/Ember/Vue.js/..."

This may be true, but tackling small refactors like this is a great way for us to become familiar before we decide to build out an entire SPA to replace our HTML views.

I spent a few hours and threw together [a prototype of this table using React.js](https://github.com/Cadasta/cadasta-platform/compare/enhancement/%231465...enhancement/%231465-react).

It's not complete, however I got pretty close with not too much effort. In the process, it exposed a pretty bad inefficiency in our DB lookups (Cadasta/django-tutelary#36) and has highlighted some short-comings in our API (missing fields needed to render table, missing ability to order or search).  Resolving these issues and expanding our API functionality would be (or, I suppose, would _have been_) simpler than building out new `/async` endpoints and would move our product forward as opposed to accruing debt.

While it's painful, I vote to stop throwing good money after bad and instead take this as an opportunity to start building tooling that moves us forward.

If you're curious to play with the prototype, running `yarn && node node_modules/babel-cli/bin/babel --plugins transform-react-jsx -w cadasta/core/static/jsx/paginated-table.jsx -o cadasta/core/static/dist/paginated-table.js` should get you going (provided you have `yarn` installed, you can swap that out with `npm install` if you prefer).  Regarding dependencies, this adds `react-dom.min.js` (38.8KB), `react.min.js` (7.8KB), and `paginated-table.js` (8.3KB, however this is unminified), for a total of 54.9KB. On the upside, it removes the dependency for `datatables.min.js` and `datatables.conditionalPaging.js` which is a combined 72.5KB.  If we end up going this route, the dependencies would probably be hosted in-house and we would need to add the build step to our deployment scripts.

**Comment by alukach** [comment]
I guess we should also validate why we are even doing this via JavaScript. Consider the tables on the Django Admin, those are done with standard HTML and using GET parameters to encode queries and sorting in the URL. Should this be our way forward as it fits more with the architecture of the rest of the site?

(I'm not actually suggesting that we should be using only HTML, but we should at least all agree that a AJAX-based solution is preferable if that's the course we are taking and then apply that reasoning moving forward.)

**Comment by oliverroick** [comment]
Closing this now, as we looking into other solutions to the problem.
